File: JavaScript/src/Internals/Functions/MathSum/Function_COUNT.js
import { Function_N } from '../Function_N.js';
import { Operand } from '../../../Operand.js';
import { FunctionUtil } from '../FunctionUtil.js';

class Function_COUNT extends Function_N {
    get Name() {
        return "Count";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args = [];
        let error = this.tryEvaluateAll(engine, tempParameter, args);
        if (error != null) {
            return error;
        }

        // 对齐 C# FlattenToList(List<Operand>, List<Operand>):不过滤类型,文本/null 也计数
        let count = 0;
        function flatten(item) {
            if (item.IsArray) {
                for (const sub of item.ArrayValue) {
                    flatten(sub);
                }
            } else if (item.IsJson) {
                let array = item.ToArray(null);
                if (array.IsError) {
                    count = -1;
                    return;
                }
                for (const sub of array.ArrayValue) {
                    flatten(sub);
                }
            } else {
                count++;
            }
        }
        for (let item of args) {
            flatten(item);
        }
        if (count < 0) {
            return this.functionError();
        }
        return Operand.Create(count);
    }
}

export { Function_COUNT };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathSum/Function_COUNT.java
package toolgood.algorithm.internals.functions.mathSum;

import java.util.List;
import java.util.ArrayList;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.FunctionUtil;
import toolgood.algorithm.internals.functions.Function_N;

final class Function_COUNT extends Function_N {
    public Function_COUNT(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length < 1) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires at least 1 parameter.");
        }
    }

    @Override
    public String Name() { return "Count"; }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        List<Operand> args = new ArrayList<>(funcs.length);
        Operand error = TryEvaluateAll(engine, tempParameter, args);
        if (error != null) { return error; }

        List<Operand> list = new ArrayList<>();
        boolean o = FunctionUtil.FlattenToList(args, list);
        if (!o) { return FunctionError(); }
        return Operand.Create(list.size());
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        if (funcs.length == 1) {
            funcs[0].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        } else {
            for (int i = 0; i < funcs.length; i++) {
                funcs[i].GetParameterTypes(noneEngine, result, OperandType.NONE);
            }
        }
    }
}
